### ISSUE TITLE
VSCode analyses `import { sqlite3 } from "sqlite3"` differently, and wrongly, from any other import in JS mode.

### ISSUE THREAD

**Issue by Pomax** [microsoft/TypeScript#45160]
Does this issue occur when all extensions are disabled?: Yes/No

Versions:

![image](https://user-images.githubusercontent.com/177243/126046994-a80f01ec-155b-45ca-8c1f-8c298bce112c.png)

Problem:

![image](https://user-images.githubusercontent.com/177243/126046988-4e9e9b77-bfaf-4601-8f0c-472579e34ae8.png)

the import is very clearly used, yet VSCode refuses to acknowledge it is. No matter how immediately I use it, VSCode keeps complaining:

![image](https://user-images.githubusercontent.com/177243/126047022-48d1f480-53c7-408b-837f-712ec914d9c8.png)

**Comment by wangweixuan** [comment]
I refer to TypeScript because *VS Code internally uses TypeScript compiler to analyse JavaScript & TypeScript files*. TypeScript is used for type checking, intellisense, etc. *Unused variable checking* is also supported by TypeScript, not by syntax parser.

As I said earlier, when you hover over the import statement, you can see the TypeScript definition file that is used by VS Code:
![import statement points to TypeScript definition file](https://user-images.githubusercontent.com/24620923/126737459-b0578ad0-ce21-4523-85f7-91c540920eb0.png)

Then delete that file, and you can see the message is gone:
![message is gone](https://user-images.githubusercontent.com/24620923/126737805-95e8fe3c-f3b8-4090-aa46-15506411c453.png)
which proves the "unused" message is from TypeScript.

Additionally, these features are powered by a builtin extension called `TypeScript and JavaScript Language Features`. Another extension called `JavaScript Language Basics` is responsible for parsing JavaScript syntax. You can try to disable the former extension and reload window, and observe that the message is gone (and all intellisense features are gone too), while syntax highlighting remains.

After all, search `sqlite3` [in VS Code codebase](https://github.com/microsoft/vscode/search?q=sqlite3&type=code) or [in TypeScript codebase](https://github.com/microsoft/TypeScript/search?q=sqlite3&type=code) yields no relevant results. It's highly unlikely that they are doing anything wrong specifically to `sqlite3`.

Finally, I suggest you change your code. As Node.JS reports, there is no named export called `sqlite3` in that module: 
![node reports error](https://user-images.githubusercontent.com/24620923/126738482-30f66799-05d0-4a30-96c7-f7b3aa543e88.png)

**Comment by Pomax** [comment]
And what VSCode uses internally should be irrelevant for user-facing behaviour: if a user uses modern JS syntax to import a namespace from a module, VSCode _should not_ flag that as erroneous, if the syntax is 100% correct and valid.

**Comment by wangweixuan** [comment]
Unused variables is not a syntax problem, but a semantic problem. And that code is not semantically correct, as Node.JS reports runtime error.

However, I agree that there is space for improvement. The error message should indicate that you are importing the wrong thing, instead of saying the import is unused.

**Comment by Pomax** [comment]
Please stop talking about node: this isn't about node, this is about pure JS.

**Comment by RyanCavanaugh** [comment]
TL;DR - all imports in JS files should have their 'value' bit set, because in this case we're getting confused and thinking it's a type-only import, which it can't be, thus we're missing the in-fact usage of `sqlite3`.

**Comment by Lemmingh** [comment]
VS Code has a setting called "Disable Automatic Type Acquisition" (`typescript.disableAutomaticTypeAcquisition`).

TypeScript has an option section called `typeAcquisition`.

IMO, that is enough to close this issue.

---

I don't understand the "bug" here.

> all imports in JS files should have their 'value' bit set
>
> we're getting confused and thinking it's a type-only import

Do [you](https://github.com/microsoft/TypeScript/issues/45160#issuecomment-885784238) mean blindly regarding every imported identifier as valid, so that related linting together with type checking will never apply to JavaScript?

But I think a more reasonable behavior would be filtering definitions when in `js` files, as TypeScript interfaces cannot be imported by JavaScript or similar languages like Flow. Then,

```javascript
import { sqlite3 } from "sqlite3";
```

should by default give:

```log
Module '"sqlite3"' has no exported member 'sqlite3'. ts(2305)

Please check https://www.npmjs.com/package/sqlite3

If you're not using this package, please check your Type Acquisition options.
```

---

The discussion started from "unused variable checking", a kind of **linting**, which involves semantics inference.

**Syntactically correct thing is never supposed to be correct.**

For example, the code below is perfectly correct in syntax, but totally wrong in semantics.

```javascript
"use strict";

let o = Object.freeze(Object.create(null));
o.a = 2;
```

It will be parsed and executed smoothly, but will eventually throw a runtime error:

```log
TypeError: Cannot add property, object is not extensible
```

Even worse, it will fail silently, if you remove the `"use strict"` statement.

Linting is recognizing pitfalls and showing suggestions to protect dev from creating things that make no sense at **runtime**.

From ESLint:

> identifying and reporting on patterns found in code, with the goal of making code more consistent and avoiding bugs.

---

The semantics of imported identifiers exactly comes from module resolution.

> [VS Code somehow knows what `sqlite3` is, even on a new file that has never been saved](https://github.com/microsoft/TypeScript/issues/45160#issuecomment-884517204)

AFAIK, VS Code (TypeScript) is asked to [guess context](https://code.visualstudio.com/docs/nodejs/working-with-javascript#_typings-and-automatic-type-acquisition) by default when working on JavaScript. Here, it assumes `"sqlite3"` to be the popular package [`sqlite3`](https://www.npmjs.com/package/sqlite3).

[You say the behavior is not desired](https://github.com/microsoft/TypeScript/issues/45160#issuecomment-885251252):

> VS Code makes assumptions about what the "sqlite3" module is when it shouldn't, because in the absence of any information other than "this is JS code"

Why? **Given that Automatic Type Acquisition is enabled.**

> [it should not make any assumptions](https://github.com/microsoft/TypeScript/issues/45160#issuecomment-885710547)

If you don't need Automatic Type Acquisition, disable it.

**Comment by Pomax** [comment]
VS Code has a "typescript" filetype. Having automatic type acquisition kick in for those makes perfect sense. Having it kick in for anything does not, so yeah: if I'm working in plain JS, VS code automatically grabbing sqlite3 type information for a project that _doesn't use that_ is still a bug and should be fixed.

**Comment by Lemmingh** [comment]
Don't you read the docs?

Automatic Type Acquisition is a well-known feature designed for JavaScript. It's the critical component behind VS Code's out-of-the-box zero-configuration support for JavaScript development. The first section of the docs gives a clear explanation.

**Comment by Pomax** [comment]
No, of course I don't read the docs if they can only be read when you already know which settings.json key is involved.  Documentation is only as good as it's discoverable, which is _super_ not the case here.

Pick sane defaults, let users override those. Not the other way around.

**Comment by Lemmingh** [comment]
[#884517204]: https://github.com/microsoft/TypeScript/issues/45160#issuecomment-884517204

[#885251252]: https://github.com/microsoft/TypeScript/issues/45160#issuecomment-885251252

[#885404969]: https://github.com/microsoft/TypeScript/issues/45160#issuecomment-885404969

[#885710547]: https://github.com/microsoft/TypeScript/issues/45160#issuecomment-885710547

[#913079035]: https://github.com/microsoft/TypeScript/issues/45160#issuecomment-913079035

[#913825131]: https://github.com/microsoft/TypeScript/issues/45160#issuecomment-913825131

I can conclude two points now.

## You refuse any guide and hardly read documentation

> If there is an expectation that people "read the docs" before they download and use VS Code, there is a wild mismatch between expectations and how people actually use VS Code.

[This][#913825131] is equal to say

> I do never read operation regulations even a short note, before entering a laboratory. It's OK if I die there.

**Users of professional software are expected to read docs as soon as they encounter trouble.** What did you see after clicking the "Download" button? [Thanks for downloading! Getting Started](https://code.visualstudio.com/docs). What did you see when opening VS Code for the first time? [The "Welcome" page](https://code.visualstudio.com/docs/getstarted/tips-and-tricks). They anticipated problems, and offered help in advance.

**A code editor / IDE is a kind of professional software.** Professional software is not toy, game, or other entertainment software which can be understood by just playing it. (Well, some toys also rely on correct understanding of their instructions.)

**Professional software can pretend to be simple, but actually is complex and requires professional knowledge.** Documentation is still the definitive textbook. A "walk-through" or "interactive playground" can only cover a small aspect. No one cannot grasp professional knowledge by just watching introductions and playing games, no matter you say how it's enjoyable, intuitive, or whatever.

---

> I don't read the docs if they can only be read when you already know which `settings.json` key or "name of the feature" is involved.

I don't understand. [Could you please elaborate on it?][#913825131]

The tutorial tells you that there's "IntelliSense". The concept is so important that it's included in a short walk-through.

![VS Code 1.60.0 - Interactive Playground](https://user-images.githubusercontent.com/43396014/132371895-73359600-dea4-4ce4-8bc4-f1050e5c6b2c.png)

The complete [guide](https://code.visualstudio.com/docs/nodejs/working-with-javascript) tells you that there's "Automatic Type Acquisition". It begins with "IntelliSense", with "Typings and Automatic Type Acquisition" immediately below.

> Type declaration files are automatically downloaded and managed by Visual Studio Code for packages listed in your project's `package.json` or that **you import into a JavaScript file**.

**They tried to inform you.**

But you were confident for ignorance, refused any help, and didn't consult dictionaries, believing you could guess the meaning of terminology. Right?

---

> [This is normal ES module syntax.][#884517204]
>
> VS Code should not know what `sqlite3` is.

> [We're talking about pure, spec-compliant ECMA-262 ECMAScript module syntax.][#885251252]

> [This isn't about Node.js. This is about pure, spec-compliant JavaScript analysis performed on string data without ever being saved to a real file and without ever running it.][#885710547]
>
> VS Code cannot know what "`sqlite3`" is yet, since there is no project backing this file.

> [If there is no project with a `package.json`, then there is no typing to guess at, and no type guessing should happen until the type system can confirm that the library name in the source is the one that belongs to the type hints it knows about.][#913825131]

I understand your idea. You don't have to repeat.

I'm not going to duplicate my comments, or the design of the TypeScript compiler and VS Code. Just add one thing:

**The language specs are generic, but the industry lives with implementations.** Module specifiers are not opaque at implementation level. Loaders establish conventions and specs to define semantics of specifiers. Your code will eventually be deployed to a concrete environment. Thus, it's safe to infer dependencies by convention.

---

Put above statements together. **You contradict yourself.**

On the one hand, you ask an editor to work, without configuration or leaning.

On the other hand, you ask the editor not to work, until you learn and configure it.

### Your lesson

Stop imagining and **learn first**.

## TypeScript's messages are obscure

As mentioned above

* [The error message should indicate that you are importing the wrong thing, instead of saying the import is unused][#885404969]

* [A more reasonable behavior would be filtering definitions][#913079035]

For such statements:

```javascript
import { sqlite3 } from "sqlite3";
sqlite3.verbose();
```

The compiler indeed suggests something goes wrong, but it's not concise, especially from the perspective of JavaScript:

```log
'sqlite3' is declared but its value is never read. ts(6133)

'sqlite3' only refers to a type, but is being used as a value here. ts(2693)
```

It can be changed to an obvious `ts(2305)` error, with additional message providing clues to the docs:

```log
Module '"sqlite3"' has no exported member 'sqlite3'. ts(2305)

Please check https://www.npmjs.com/package/sqlite3

If you're not using this package, please check your Type Acquisition options.
```

### TypeScript's lesson

Give clear diagnostic information with **links to the docs**.

**Comment by Pomax** [comment]
Then I guess we're done here.
